src/org/kms/core/MKRequester.java

```diff
@@ -95,13 +99,36 @@ public String getMasterKey() {
 			
 		} catch (ErrorResponse error) {
 			if (error.getErroCode() == 404){
+				String uid = getUserIdFromRequestMK();
 				try {
-					master_key = generateRawMasterKey();
-					String encrypted_master_key = crypto.encryptKey(master_key, crypto.getGlobalKey());
-					if (putMasterKeyToBackend(encrypted_master_key)){
-						decrypted_master_key = master_key;
+					if (!KMSUtils.putWithLock(uid)) {
+						while (KMSUtils.getMap().contains(uid)){
+							Thread.sleep(50);
+						}
```

Comment:
what is the logic behind this loop ? KMSUtils.putWithLock(uid) will return false only when uid is present in the set.